README.md
# Wayback Restore JS

> This package is still experimental.

A website restoration tool written for Node Js.

This package allows you to restore a website from web.archive.org. It was developed for Node JS and can be used in any Electron application.

## Install

```
npm install wayback-restore
```

## Usage

```JavaScript
var Wayback = require("wayback-restore");

Wayback.restore(options);
```

## API

### restore(options)

#### options

##### url - string

URL to restore. Ex. https://web.archive.org/web/20150801040409/http://example.com/

If you use `url` then you do not need to use `timestamp` and `domain`.

##### timestamp - string

Timestamp to restore.

Ex. 20150801040409

##### domain - string

Domain to restore

##### directory - string

(Default: restore) Directory to output into.

##### max_pages - integer

(default: no limit), Maximum number of pages to download. Leave empty for no limit.

##### links - boolean

(default: true) Set to true to download links found on the page.

##### assets - boolean

(default: true) Set to true to download CSS, JS, images.

##### concurrency - integer

(default: 1): Number of downloads to process at once.

**_Warning: Setting this value too high might get you blocked from web.archive.org._**

##### log - boolean

(default: false) Set to true to enable logging to a log file.

##### logFile - string

(default: restore.log) Name of the log file to write. It will be written to `options.directory`.

#### Methods

##### start

##### stop

##### pause

##### resume

#### Events

The following events are emitted.

##### start(callback)

Fired when restoring starts.

```
.on("start", function() {
    console.log("[STARTED USING]:", this.settings);
})
```

##### restoring - (asset)

When a file begins restoring.

```
.on("restoring", function(asset) {
   console.log("[RESTORING]", asset.original_url);
})
```

##### restored - (asset)

Fired when a file has been downloaded.

```
.on("restored", function(asset) {
   console.log("[RESTORED]", asset.original_url);
})
```

##### cdxquery

```
.on("cdxquery", function(cdx) {
  console.log("Snapshots Found: ", cdx.size);
})
```

##### completed - (results)

When the restore process has completed.

```
.on("completed", function(results) {
    console.log("restoration has completed");
    console.log("url: ", results.url);
    console.log("domain: ", results.domain);
    console.log("timestamp: ", results.timestamp);
    console.log("directory: ", results.directory);
    //console.log("first file: ", results.first_file);
    //console.log("started: ", results.started);
    //console.log("ended: ", results.ended);
    console.log("restored: ", results.restored_count);
    console.log("failed: ", results.failed_count);
    console.log("Runtime:", results.runtime_hms);
})
```

### downloader(options)

#### options

##### url - string

A snapshot URL to download.

```JavaScript
Wayback.download({
    url: 'http://web.archive.org/web/20150531/http://www.example.com'
});
```

##### domain - string

A domain to download from.

##### from - string

Only files on or after timestamp supplied (ie. 20150801231334).

Can also be 20150801

##### to - string

Only files on or before timestamp supplied (ie. 20150801231334).

Can also be 20150801

##### limit - integer (default: 0)

limit number of files to download.

0 = no limit.

##### exact_url - boolean (default: false)

Downloads only the url provided not full site.

##### list - boolean (default: false)

Doesn't download any files.

##### concurrency - integer (default: 1)

Number of files to download at the same time.

##### only - string|RegEx (defalt: '')
    
Only include files matching this filter.

##### exclude - string|RegEx (defalt: '')
    
Excludes files matching this filter.

##### directory - string (defalt: '.')

Directory to output into.

#### Methods

See Wayback.restore methods.

### Asset (object)

An `Asset` is an object that gets downloaded or restured and is returned by various events.

#### Asset.properties

  // CDX urlkey
  this.key = null;

  // the url to restore
  this.original_url = "";

  // path to local file
  this.restored_file = "";

  this.timestamp = "";

  // restored | failed | unarchived
  this.status = RESTORE_STATUS.EMPTY;

  // mimetype: html, image, css, js, based on wayback types
  this.mimetype = "";

#### Asset.methods

##### Asset.getSnapShot(flag)

Set to `true` to fetch file in its original state, without any processing by the Wayback Machine or waybackpack. You will likely want this to always be `true`.


## Examples

```JavaScript
var restore = Wayback.restore({
    url:
        "http://web.archive.org/web/20150531/http://example.com"
});
```

```JavaScript
var restore = Wayback.restore({
    domain: 'example.com',
    timestamp: "20150531"
});
```

```JavaScript
var restore = Wayback.restore('http://web.archive.org/web/20150531/http://example.com');
```


```JavaScript
Wayback.downloader({
  url: 'https://trufish.org/',
  from: '20181001',
  to: '20201031',
  list: true,
  concurrency: 10,
  exact_url: false,
  exclude: /.(gif|jpg|jpeg|png|svg)$/i
})
  .on('completed', function (results) {
    console.log('completed');
    console.log(results);
  })
  .start((asset) => {
    console.log('Asset', asset.getSnapshotUrl());
  });
```

## TODO

- Improve documentation
- Improve this modules API
- Create a CLI. Maybe as a separate module?

## Need a GUI Application?

Checkout [Restorizor](https://www.restorizor.com) a Wayback Machine download application built using Electron and powered by this very same wayback-restore.js module.


CHANGELOG.md

# Changelog

## 0.0.9
- Changed restore process to find snapshots based on timestamp YYYMMDD instead of including hour, minutes, seconds
- Created this CHANGELOG.md file


restore.js
/**
 * This file only exists for testing restore/downloader processes by executing
 * `npm run restore`.
 *
 * @TODO: replace this with a CLI utility.
 */

var Wayback = require('./wayback-restore');

/**
 * Test URLS
 */
//const url = "https://web.archive.org/web/20150531/http://www.cashpropertysolutions.co.uk",
//const url = 'https://web.archive.org/web/20170204050649/http://www.androidfantasy.com/',
//const url = "https://web.archive.org/web/20150801040409/http://acbaw.com/",
//const url = "http://web.archive.org/web/20091125054126/http://www.ulcinjtoday.com/",
//const url = "https://web.archive.org/web/20190114224925/http://www.tennisballmachinereviews.org/",
//const url = "https://web.archive.org/web/20200602050304/https://www.fancytextguru.com/",
//const url = "https://web.archive.org/web/20190424225217/http://remont-k.com/",
//const url = "https://web.archive.org/web/20181029143918/https://trufish.org/";
//const url = "https://web.archive.org/web/20150424013550/http://www.kbect.com/",
//const url = "https://web.archive.org/web/20170923120200/http://careersters.net/",
//const url ="https://web.archive.org/web/20160404213504/http://claridadesdemichoacan.com",

// @TODO: this doesn't not work because it's an IP
//const url = "https://web.archive.org/web/20200612100421/http://198.96.92.14/",
// @NOTE: This is not a valid URL and does not restore
//const url = "https://web.archive.org/web/20181029143918/trufish.org",

/*
Wayback.restore({
  url: 'http://web.archive.org/web/20150531/http://www.cashpropertysolutions.co.uk'
});

Wayback.restore({
  domain: 'cashpropertysolutions.co.uk',
  timestamp: '20150531'
});

Wayback.restore('http://web.archive.org/web/20150531/http://www.cashpropertysolutions.co.uk');
*/

/*
Wayback.restore({
  directory: '~/testrestore/restores/',
  url: 'http://web.archive.org/web/20150531/http://www.cashpropertysolutions.co.uk',
  //  url: 'https://web.archive.org/web/20150202214825/http://dollarseoclub.com/#!#',
  concurrency: 1,
  max_pages: 50,
  links: true
})
  .on('completed', function (results) {
    console.log('restoration has completed');
    console.log('url: ', results.url);
    console.log('domain: ', results.domain);
    console.log('timestamp: ', results.timestamp);
    console.log('directory: ', results.directory);
    console.log('started: ', results.started);
    console.log('ended: ', results.ended);
    console.log('restored: ', results.restored_count);
    console.log('failed: ', results.failed_count);
    console.log('Runtime:', results.runtime_hms);
  })
  .on('restoring', function (asset) {
    console.log('[RESTORING]', asset.original_url);
  })
  .on('restored', function (asset) {
    console.log('[RESTORED]', asset.original_url);
  })
  .on('start', function () {
    console.log('[STARTED USING]:', this.settings);
  })
  .on('cdxquery', function (cdx) {
    console.log('Snapshots Found: ', cdx.size);
  })
  .start();
*/

/**
 * Downloader Examples.
 */

Wayback.downloader({
  concurrency: 5,
  directory: '/home/nick/testrestore',
  domain: 'trufish.org',
  exclude: '',
  limit: 10,
  notify_on_finish: true,
  //only: '/.(gif|jpg|jpeg|png)$/i',
  //only: '.(gif|jpg|jpeg|png)$',
  //from: '2017',
  //to: '2018',

  //from: '',
  //to: '',

  list: true

  //url: 'https://trufish.org/',
  //from: '20181001',
  //to: '20201031',
  //list: false,
  //limit: 8,
  //concurrency: 10,
  //exact_url: false,
  //only: 'https://trufish.org/wp-content/themes/aspire-pro/images/bg-1.jpg',
  //exclude: /.(gif|jpg|jpeg|png|svg)$/i
})
  .on('completed', function (results) {
    console.log('completed');
    console.log(results);
  })
  .start((record) => {
    console.log('Asset', record.getSnapshotUrl());
  });
